src/main/java/io/smartcat/ranger/core/CircularRangeValue.java

```diff
@@ -0,0 +1,91 @@
+package io.smartcat.ranger.core;
+
+/**
+ * Base circular range value class.Generates values within specified <code>range</code> from the beginning to the end
+ * with specified <code>increment</code>.
+ *
+ * @param <T> Type this value would evaluate to.
+ */
+public abstract class CircularRangeValue<T extends Number & Comparable<T>> extends Value<T> {
+
+    /**
+     * Range.
+     */
+    protected final Range<T> range;
+
+    /**
+     * Increment.
+     */
+    protected final T increment;
+
+    /**
+     * Current value within range.
+     */
+    protected T currentValue;
+
+    private boolean firstEvaluation = true;
+
+    /**
+     * Constructs circular range value with specified <code>range</code> and <code>increment</code>.
+     *
+     * @param range Range.
+     * @param increment Increment.
+     */
+    public CircularRangeValue(Range<T> range, T increment) {
+        if (range == null) {
+            throw new IllegalArgumentException("Range cannot be null.");
+        }
+        if (increment == null || increment.compareTo(zero()) == 0) {
+            throw new IllegalArgumentException("Increment cannot be null nor zero.");
+        }
+        if (range.isEmpty()) {
+            throw new IllegalArgumentException("Range cannot be empty.");
+        }
+        if (range.isIncreasing() && increment.compareTo(zero()) < 0) {
+            throw new IllegalArgumentException("If range is increasing, increment must be positive.");
+        }
+        if (range.isDecreasing() && increment.compareTo(zero()) > 0) {
+            throw new IllegalArgumentException("If range is decreasing, increment must be negative.");
+        }
+        this.range = range;
+        this.increment = increment;
+        this.currentValue = range.getBeginning();
+        if (isIncrementGreaterThanRangeSize()) {
```

Comment:
Well, then beats the purpose of circular value. Then I would need to modify iteration code to support this, since now I reset to beginning when value reaches over the end.